src/mixins/helpers.js

```diff
@@ -190,17 +190,27 @@ var helpers = {
       targetLeft = currentLeft;
     }
 
-    if(this.props.afterChange!==null){
-      this.props.afterChange(currentSlide);
+    if (this.props.beforeChange) {
+      this.props.beforeChange(currentSlide);
     }
 
-    var callback = function() {
-        this.setState({
-          animating: false,
-          trackStyle: this.getCSS(currentLeft),
-          swipeLeft: null
-        });
-    }.bind(this);
+    var nextStateChanges = {
+      animating: false,
+      trackStyle: this.getCSS(currentLeft),
+      swipeLeft: null
+    };
+
+    var callback1 = () => {
+      this.setState(nextStateChanges);
+    };
+
+    var callback2 = () => {
+      this.setState(nextStateChanges);
+
+      if (this.props.afterChange) {
+        this.props.afterChange(currentSlide);
+      }
+    };
```

Comment:
addEndEventListener, and setTimeout are for same purpose. We should not give different callback to them.
setTimeout is backup for addEndEventerListener.

What if you call the afterChange function in callback and pass the same callback to both. Will it work ?